nixos/modules/services/backup/bareos.nix

```diff
@@ -0,0 +1,292 @@
+        Password = "${fd_cfg.director.password}"
+        Address = ${fd_cfg.director.address}
+        ${optionalString fd_cfg.tls.enable ''
+          TLS Enable = ${boolSwitch fd_cfg.tls.enable}
+          TLS Require = ${boolSwitch fd_cfg.tls.require}
+          TLS Verify Peer = ${boolSwitch fd_cfg.tls.verify}
+          TLS ALLOWED CN = ${fd_cfg.tls.allowed}
+          TLS CA Certificate File = ${fd_cfg.tls.ca}
+          TLS Certificate = ${fd_cfg.tls.cert}
+          TLS Key = ${fd_cfg.tls.key}
+          TLS DH File = ${fd_cfg.tls.dh} ''}
+        ${fd_cfg.extraDirectorConfig}
+      }
+
+      FileDaemon {
+        Name = ${fd_cfg.name}
+        WorkingDirectory = "${libDir}"
+        ${concatMapStrings ({ protocol, addr, ... }: ''
+          FDAddresses = { ${protocol} = { addr = ${addr} }} '') fd_cfg.listen}
+        Compatible = ${fd_cfg.compatible}
+        ${optionalString fd_cfg.tls.enable ''
+          TLS Enable = ${boolSwitch fd_cfg.tls.enable}
+          TLS Require = ${boolSwitch fd_cfg.tls.require}
+          TLS CA Certificate File = ${fd_cfg.tls.ca}
+          TLS Certificate = ${fd_cfg.tls.cert}
+          TLS Key = ${fd_cfg.tls.key} ''}
+        ${optionalString fd_cfg.pki.enable ''
+          PKI Signatures = ${boolSwitch fd_cfg.pki.signature}
+          PKI Encryption = ${boolSwitch fd_cfg.pki.encryption}
+          PKI Keypair = ${fd_cfg.pki.keypair}
+          PKI Master Key = ${fd_cfg.pki.key} ''}
+        ${fd_cfg.extraClientConfig}
+      }
+
+      Messages {
+        Name = Standard
+        director = "${fd_cfg.director.name}" = all, !skipped, !restored
+        ${fd_cfg.extraMessagesConfig}
+      }
+    '';
+
+in {
+  options = {
+    services.bareos-fd = {
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Whether to enable the Bareos File Daemon.
+        '';
+      };
+
+      name = mkOption {
+        default = "${config.networking.hostName}-fd";
+        description = ''
+          The client name that must be used by the Director when connecting.
+          Generally, it is a good idea to use a name related to the machine
+          so that error messages can be easily identified if you have multiple
+          Clients. This directive is required.
+        '';
+      };
+
+      listen = mkOption {
+        type = with types; listOf (submodule {
+          options = {
+            protocol = mkOption {
+              type = types.nullOr types.str;
+              default = null;
+              description = ''
+                IP type, IPv4 or IPv6.
+              '';
+            };
+            addr = mkOption {
+              type = types.nullOr types.str;
+              default = null;
+              description = ''
+                IP address client should listen on.
+              '';
+            };
+          };
+        });
+      };
+
+      compatible = mkOption {
+         default = "yes";
```

Comment:
Already using a function for bool mapping, can extend it to other true/false, yes/no options